### PR TITLE
🧹feat: Making course detail screen 

### PR DESCRIPTION
**Pull request by hungmai2411**
## Key Changes:

**New Feature:**
  * Introduced a new Course Detail Screen in the UI using Jetpack Compose.

### CODE REVIEW

File: feat/src/main/java/com/example/feat/course_detail/CourseDetailScreen.kt [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+@Composable
+fun CourseDetailScreen(
+    courseId: String,
+    courseDetailViewModel: CourseDetailViewModel = hiltViewModel()
+) {
+    val lazyScrollState = rememberLazyListState()
+    val uiState = courseDetailViewModel.uiState.collectAsState()
+    LaunchedEffect(Unit) {
+        courseDetailViewModel.onEvent(CourseDetailUiEvent.FetchCourseDetail(courseId))
+    }
```
Comment (nguyenminhhung2011) [inline feat/src/main/java/com/example/feat/course_detail/CourseDetailScreen.kt:56]: The `LaunchedEffect` with `Unit` key will run only once when the `CourseDetailScreen` composable enters the composition. If you need to refetch course details based on `courseId`, consider using `LaunchedEffect(courseId)` instead.

```diff
-LaunchedEffect(Unit) {
+LaunchedEffect(courseId) {
    courseDetailViewModel.onEvent(CourseDetailUiEvent.FetchCourseDetail(courseId))
```

File: feat/src/main/java/com/example/feat/course_detail/CourseDetailScreen.kt [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+            if (uiState.value.courseDetail != null) {
+                CourseDetailToolbar(
+                    lazyScrollState = lazyScrollState,
+                    title = uiState.value.courseDetail?.name ?: "",
+                    backgroundImageUrl = uiState.value.courseDetail?.imageUrl ?: "",
+                )
+            }
+        }
+    ) { paddingValues ->
+        Column(
```
Comment (nguyenminhhung2011) [inline feat/src/main/java/com/example/feat/course_detail/CourseDetailScreen.kt:73]: Why wrap the Column outside like this, couldn't just have the LazyColumn outside instead?

File: feat/src/main/java/com/example/feat/course_detail/components/CourseDetailToolbar.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    ) {
+        if (backgroundImageUrl != null && isExpanded) {
+            AsyncImage(
+                model = backgroundImageUrl,
+                contentDescription = null,
+                contentScale = ContentScale.FillBounds,
+                modifier = Modifier
+                    .layoutId("backgroundImage")
+            )
+        } else {
```
Comment (nguyenminhhung2011) [inline feat/src/main/java/com/example/feat/course_detail/components/CourseDetailToolbar.kt:83]: Consider checking if the URL is a valid and accessible URL before attempting to load the image, to improve error handling and avoid crashes.

```diff
- if (backgroundImageUrl != null && isExpanded) {
+ if (!backgroundImageUrl.isNullOrBlank() && isExpanded) {
```

File: feat/src/main/java/com/example/feat/course_detail/components/CourseDetailToolbar.kt [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    lazyScrollState: LazyListState,
+    backgroundImageUrl: String? = null,
+    title: String? = null
+) {
+    val context = LocalContext.current
+    val configuration = LocalConfiguration.current
+    val screenHeight = configuration.screenHeightDp
+
+    val motionScene = remember {
+        context.resources.openRawResource(R.raw.collapse_toolbar).readBytes().decodeToString()
```
Comment (nguyenminhhung2011) [inline feat/src/main/java/com/example/feat/course_detail/components/CourseDetailToolbar.kt:48]: Using `readBytes().decodeToString()` can potentially lead to issues with large files and memory usage. Consider using `BufferedReader` for better performance with large resources.

```diff
- context.resources.openRawResource(R.raw.collapse_toolbar).readBytes().decodeToString()
+ context.resources.openRawResource(R.raw.collapse_toolbar).bufferedReader().use { it.readText() }
```

Explanation
* readBytes().decodeToString(): This approach reads the entire file into memory as a byte array, which can lead to high memory consumption for large files.
* bufferedReader().use { it.readText() }: This method reads the file line by line (or in chunks) using a buffer, which is more efficient and reduces memory overhead.

**Review by nguyenminhhung2011 — CHANGES_REQUESTED** [review]
@hungmai2411

File: feat/src/main/java/com/example/feat/course_detail/controller/CourseDetailViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            is CourseDetailUiEvent.FetchCourseDetail -> fetchCourseDetail(event.id)
+        }
+    }
+
+    private suspend fun fetchCourseDetail(id: String) {
+        setUiState { copy(isLoading = true) }
+        viewModelScope.launch {
+            courseDetailUseCase.fetchCourseDetail(id).collect { either ->
+                if (either.isRight()) {
+                    setUiState { copy(courseDetail = either.rightValue()) }
```
Comment (nguyenminhhung2011) [inline feat/src/main/java/com/example/feat/course_detail/controller/CourseDetailViewModel.kt:25]: Write an additional state to handle the failed data, please.

File: feat/src/main/java/com/example/feat/course_detail/components/CourseDetailToolbar.kt [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                imageVector = Icons.AutoMirrored.Filled.ArrowBack,
+                contentDescription = "Back Button",
+                tint = Color.White
+            )
+        }
+
+        Text(
+            text = title ?: "",
+            modifier = Modifier.layoutId("title"),
+            style = if (!isExpanded) DesignSystem.TITLE_SMALL_STYLE.copy(
```
Comment (nguyenminhhung2011) [inline feat/src/main/java/com/example/feat/course_detail/components/CourseDetailToolbar.kt:105]: Why check the negation of a condition? It violates clean code.

**Review by nguyenminhhung2011 — CHANGES_REQUESTED** [review]

File: feat/src/main/java/com/example/feat/course_detail/controller/CourseDetailViewModel.kt (hunk 5 above)
Comment (hungmai2411) [reply feat/src/main/java/com/example/feat/course_detail/controller/CourseDetailViewModel.kt:25]: here i want either the result is right or left => `setUiState { copy(isLoading = false) }`

**Review by nguyenminhhung2011 — APPROVED** [review]
🎉 LGTM

**Pull request merged**